### PR TITLE
Cache g.Symbol.toSemantic

### PR DESCRIPTION
**Pull request by xeno-by**

### CODE REVIEW

File: scalameta/semanticdb-scalac-core/src/main/scala/scala/meta/internal/semanticdb/SymbolOps.scala [hunk 1]
```diff
@@ -1,84 +1,96 @@
 package scala.meta.internal
 package semanticdb
 
+import java.util.HashMap
 import scala.{meta => m}
 import scala.meta.internal.inputs._
 
 trait SymbolOps { self: DatabaseOps =>
 
+  private lazy val symbolCache = new HashMap[g.Symbol, m.Symbol]
```
Comment (olafurpg) [inline scalameta/semanticdb-scalac-core/src/main/scala/scala/meta/internal/semanticdb/SymbolOps.scala]: Can we maybe do this change without adding a 2 space indentation on the whole method? The diff makes it difficult to see the difference

File: scalameta/semanticdb-scalac-core/src/main/scala/scala/meta/internal/semanticdb/SymbolOps.scala [hunk 2]
```diff
@@ -1,84 +1,96 @@
...
+          else m.Signature.Term(name(sym))
+        }
+        m.Symbol.Global(owner, signature)
+      }
+      val msym = symbolCache.get(sym)
+      if (msym != null) {
+        msym
+      } else {
+        val msym = uncached(sym)
+        symbolCache.put(sym, msym)
```
Comment (olafurpg) [inline scalameta/semanticdb-scalac-core/src/main/scala/scala/meta/internal/semanticdb/SymbolOps.scala]: Is this cache cleared anywhere?

File: bench/README.md [hunk 3]
```diff
@@ -38,6 +38,7 @@ Other Scalameta configurations progressively enable more and more features:
 Benchmark                      Mode    Cnt Score
 QuickScalacBaseline            sample  35  1595.274 ± 15.263  ms/op
 QuickScalacRangepos            sample  35  1703.097 ± 25.044  ms/op  (6% overhead)
+QuickScalametaOpt              sample  15  1934.902 ± 29.232  ms/op  (21% overhead)
```
Comment (olafurpg) [inline bench/README.md]: Wow, is that a speedup from 42% to 21% overhead?

Comment (xeno-by) [reply bench/README.md]: It's just 25% to 21% :)

File: scalameta/semanticdb-scalac-core/src/main/scala/scala/meta/internal/semanticdb/SymbolOps.scala (hunk 2 above)
Comment (xeno-by) [reply scalameta/semanticdb-scalac-core/src/main/scala/scala/meta/internal/semanticdb/SymbolOps.scala]: Good point.

File: scalameta/semanticdb-scalac-core/src/main/scala/scala/meta/internal/semanticdb/SymbolOps.scala (hunk 1 above)
Comment (xeno-by) [reply scalameta/semanticdb-scalac-core/src/main/scala/scala/meta/internal/semanticdb/SymbolOps.scala]: Would you like to have a separate commit that just changes indentation?

File: scalameta/semanticdb-scalac-core/src/main/scala/scala/meta/internal/semanticdb/SymbolOps.scala [hunk 4]
```diff
@@ -1,84 +1,96 @@
...
 
+import java.util.HashMap
 import scala.{meta => m}
 import scala.meta.internal.inputs._
 
 trait SymbolOps { self: DatabaseOps =>
 
+  lazy val symbolCache = new HashMap[g.Symbol, m.Symbol]
   implicit class XtensionGSymbolMSymbol(sym: g.Symbol) {
     def toSemantic: m.Symbol = {
```
Comment (olafurpg) [inline scalameta/semanticdb-scalac-core/src/main/scala/scala/meta/internal/semanticdb/SymbolOps.scala]: Can we keep the original function unchanged except named `toSemanticUncached` and then create new `toSemantic` that uses caching? I feel it's difficult to follow the logic when the nesting is too deep. The big diff also makes it harder to review this change.

Comment (olafurpg) [reply scalameta/semanticdb-scalac-core/src/main/scala/scala/meta/internal/semanticdb/SymbolOps.scala]: Nevermind, I want to edit this method to fix #1194

**Pull request merged**